server/src/controllers/csv.js

```diff
@@ -0,0 +1,174 @@
+    const upload = multer({
+      storage,
+      fileFilter: function (req, file, cb) {
+        file.mimetype === "text/csv" ||
+          file.mimetype === "application/vnd.ms-excel"
+          ? cb(null, true)
+          : cb(null, false);
+      },
+      limits: {
+        fileSize: 5 * 1024 * 1024, //5MB
+      },
+    }).single("prospects");
+
+    // Perform upload to server
+    upload(req, res, async (err) => {
+      // Check if there were any filetype error or data storage restriction error
+      if (err) {
+        console.log("MULTER ERROR: ", err.message);
+        return res.status(401).send({ errors: [{ message: err.message }] });
+      } else {
+        try {
+          if (!req.file) {
+            return res
+              .status(401)
+              .send({ errors: [{ message: "Invalid file" }] });
+          }
+
+          const filename = req.file.filename;
+          const results = [];
+
+          // Attempt upload to cloud 
+          try {
+            const url = (
+              await axios.post(
+                process.env.GOOGLE_CLOUD_SIGN_URL,
+                {
+                  filename: filename,
+                  contentType: req.file.mimetype,
+                },
+                { headers: { "Content-type": "application/json" } }
+              )
+            ).data.url;
+
+            await axios.put(url, req.file, {
+              headers: {
+                "Content-Type": req.file.mimetype,
+              },
+            });
+          } catch (err) {
+            // console.log("Error happened");
+            console.log("Error uploading to cloud:", err);
+          }
+
+          // Create a new entry in Files table to manage all files a user has uploaded
+          await db.Files.create({
+            name: filename,
+            userId: req.currentUser.id,
+            type: req.file.mimetype,
+          });
+
+          // Parse through the csv and find what fields does it contain
+          fs.createReadStream(`./prospects/${filename}`)
+            .pipe(csv())
+            .on("data", (data) => results.push(data))
+            .on("end", () => {
+              // Check if the csv is empty
+              if (!(results.length > 0)) {
+                return res
+                  .status(400)
+                  .send({ errors: [{ message: "Empty CSV file" }] });
+              }
+
+              const properties = Object.keys(results[0]);
+
+              // Get the attributes of the Prospect Table 
+              const attributes = (Object.keys(db.sequelize.models.Prospect.rawAttributes))
+                .filter(attr => {
+                  return ['createdAt', 'updatedAt', 'status', 'id'].indexOf(attr) === -1
+                })
+              console.log(attributes)
+
+
+              // Set up a cookie equal to the file name for reference
+              req.session = {
+                ...req.session,
+                filename: filename
+              }
```

Comment:
I don't know if I want to use cookies here to do the filename.... Will this work with many users? Instead, I feel like addProspects route should take the filename from the front-end

Reply:
I thought of that too, but since I am renaming the files with uuid(), I needed a place to store filenames that were client independent and would be accessible accross that single user's session. I tried it with multiple browsers and it works fine.